Handle overlapping routes with different Regexes

rpjohnst opened the pull request:
I have two routes based on different `Regex` strings that can match the same path, but with different methods. Depending on their ordering, current `reroute` forces one to respond with a 405, because it only looks in its route table for the first matching `Regex` string.

(For context, the two routes are `OPTIONS` with `.*$` and `GET` with `/something/specific$`, where the `OPTIONS` route is a generic CORS handler for several other endpoints.)

I don't know that this is the best solution, precedence or performance-wise, but I don't have a lot of experience with other router APIs and how they solve this problem, so I'd like to see what others think here. This PR just loops over all matching `Regex`es rather than failing if the first one's method doesn't match the request.

gsquire commented:
You raise a good point with this PR. This is something that hadn't occurred to me with some simple testing. It LGTM overall but I think I want to take it a step further and try and knock out #14 too.

I think with your addition of checking each match it would be good to see if we can catch `/` matching everything by checking the length of the route path.

Perhaps something like this:
```rust
if route == "/" && uri.len() > 1 {
    continue;
}
```
Let me know if this makes sense or not.

rpjohnst commented:
Okay, I've done some research on how other routing APIs handle this problem.

The solution to the `/` problem seems to be simply anchoring the `Regex`es at the end as well as the beginning- no need to hardcode anything and it's a more general solution. One tweak we could do on top is to add `\/?` like express.js does (via `path-to-regexp`).

There are also options for disambiguating routes beyond just the order they were added- express.js takes a callback for each named parameter to validate it before accepting a route, and Conduit/Iron's `route-recognizer` orders routes by number of `*param`, `:param`, and static path elements. These both require named parameter support, but aren't really necessary either.

Performance-wise, `route-recognizer` uses a custom NFA implementation `Router<T>` that matches multiple routes at once *and* collects captures as it goes. Conduit uses a `HashMap<Method, Router<Handler>>`, while Iron uses `Router<HashMap<Method, Handler>>`.

The first solution involves no looping over candidates, but does duplicate the matcher. This is equivalent to including the method in the `Regex`es, which in our case is even worse- not only does the `RegexSet` get bigger, we also duplicate stand-alone `Regex`es for each method (which we currently do anyway).

The second has only one matcher (and in our case fewer `Regex`es), but requires a linear search to find a route with the right method. I think I'd like to go with this method, which is equivalent to this PR though a little better-organized.

Thoughts?

gsquire commented:
I am fine with the second option like you said. It also never occurred to me that you could add `/Z` to the routes as well. I wonder if this will cause any performance hits to the `RegexSet`.

Also, thanks so much for the research. I have used Iron's router when I built something with Iron and I know the routing is done quite differently than in reroute's approach.

Let me know if you need any help with this.

rpjohnst commented:
Alright, here's an update. The main change is from a `HashMap<RouteInfo, Handler>` to a `Vec<(Method, Handler)>`. I didn't wind up coalescing the regexes routes that use the same string because that doesn't preserve insertion order, and reroute already duplicates compiled regexes. Maybe we can find a solution to that later.

I also switched to the builder pattern, to cut down on the amount of redundant state in the router, and simplified the error type since it turns out `regex` as of `0.1.59` supports `RegexSet`s with 0 entries. That makes this a breaking change, not sure what you feel about that in this situation.

gsquire commented on Cargo.toml:
```diff
@@ -10,7 +10,7 @@ license = "MIT"
 [dependencies]
 hyper = { version = "0.9", default-features = false }
-regex = "0.1"
+regex = "0.1.59"
```
Is there a specific reason you used this version? I was able to compile with 0.1.80 just fine.

rpjohnst replied:
Yeah, this is the minimum required version since it supports 0-entry `RegexSet`s- Cargo will pull `0.1.80` for new projects since it's semver-compatible with `0.1.59`.

gsquire commented on src/error.rs:
```diff
@@ -1,26 +1,30 @@
 use std::{error, fmt};
+use regex;
```
Since we are just referring to the regex package's error type, can we omit the `use` statement and just refer to it as `::regex::Error`?

gsquire commented on src/lib.rs:
```diff
@@ -27,173 +21,120 @@ pub struct RouteInfo {
-    compiled_list: Vec<Regex>,
-    route_map: HashMap<RouteInfo, Box<Fn(Request, Response, Captures)>>,
+    routes: RegexSet,
+    regexes: Vec<Regex>,
```
Can we call this `patterns`?

gsquire commented on src/lib.rs:
```diff
@@ -27,173 +21,120 @@ pub struct RouteInfo {
-        self.route_list.push(anchored_pattern.clone());
-        self.route_map.insert(route_info, Box::new(handler));
-    }
+        // anchor at the start and end so routes only match exactly
```
Kind of pedantic, but I like to make my comments sentences.

gsquire commented:
Thanks so much for this PR. I will publish a new release soon after.

Closes #14
